#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""gRPC client for calling the bazi-rule-service."""

from __future__ import annotations

import json
import logging
import os
import sys
from typing import Any, Dict, List, Optional

import grpc

# ÂØºÂÖ•ÁîüÊàêÁöÑ gRPC ‰ª£Á†Å
project_root = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
sys.path.insert(0, os.path.join(project_root, "proto", "generated"))

import bazi_rule_pb2
import bazi_rule_pb2_grpc

logger = logging.getLogger(__name__)


class BaziRuleClient:
    """gRPC client for the bazi-rule-service."""

    def __init__(self, base_url: Optional[str] = None, timeout: float = 60.0) -> None:
        # base_url Ê†ºÂºè: host:port Êàñ [host]:port
        base_url = base_url or os.getenv("BAZI_RULE_SERVICE_URL", "")
        if not base_url:
            raise RuntimeError("BAZI_RULE_SERVICE_URL is not configured")
        
        # Ëß£ÊûêÂú∞ÂùÄÔºàÁßªÈô§ http:// ÂâçÁºÄÔºâ
        if base_url.startswith("http://"):
            base_url = base_url[7:]
        elif base_url.startswith("https://"):
            base_url = base_url[8:]
        
        # Â¶ÇÊûúÊ≤°ÊúâÁ´ØÂè£ÔºåÊ∑ªÂä†ÈªòËÆ§Á´ØÂè£
        if ":" not in base_url:
            base_url = f"{base_url}:9004"
        
        self.address = base_url
        self.timeout = timeout

    def match_rules(
        self,
        solar_date: str,
        solar_time: str,
        gender: str,
        rule_types: Optional[List[str]] = None,
        use_cache: bool = False,
    ) -> Dict[str, Any]:
        """ÂåπÈÖçËßÑÂàô"""
        request = bazi_rule_pb2.BaziRuleMatchRequest(
            solar_date=solar_date,
            solar_time=solar_time,
            gender=gender,
            rule_types=list(rule_types) if rule_types else [],
            use_cache=use_cache,
        )

        import datetime
        request_time = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        rule_types_str = ", ".join(rule_types) if rule_types else "ÂÖ®ÈÉ®"
        logger.info(f"[{request_time}] üîµ Ë∞ÉÁî® bazi-rule-service (gRPC): {self.address}, solar_date={solar_date}, solar_time={solar_time}, gender={gender}, rule_types=[{rule_types_str}], use_cache={use_cache}")
        logger.debug("Calling bazi-rule-service (gRPC): %s request=%s", self.address, request)

        # ËÆæÁΩÆËøûÊé•ÈÄâÈ°πÔºåÈÅøÂÖç "Too many pings" ÈîôËØØ
        # ÈÖçÁΩÆÊ∂àÊÅØÂ§ßÂ∞èÈôêÂà∂ÔºåÊîØÊåÅÂ§ßÂìçÂ∫îÔºà462Êù°ËßÑÂàôÂèØËÉΩ‰∫ßÁîüËæÉÂ§ßÁöÑÂìçÂ∫îÔºâ
        options = [
            ('grpc.keepalive_time_ms', 300000),  # 5ÂàÜÈíüÔºåÂáèÂ∞ë ping È¢ëÁéá
            ('grpc.keepalive_timeout_ms', 20000),  # 20ÁßíË∂ÖÊó∂
            ('grpc.keepalive_permit_without_calls', False),  # Ê≤°ÊúâË∞ÉÁî®Êó∂‰∏çÂèëÈÄÅ ping
            ('grpc.http2.max_pings_without_data', 2),  # ÂÖÅËÆ∏ÊúÄÂ§ö2‰∏™ ping
            ('grpc.http2.min_time_between_pings_ms', 60000),  # ping ‰πãÈó¥Ëá≥Â∞ëÈó¥Èöî60Áßí
            # Â¢ûÂä†Ê∂àÊÅØÂ§ßÂ∞èÈôêÂà∂ÔºàÈªòËÆ§4MBÔºåÂ¢ûÂä†Âà∞50MB‰ª•ÊîØÊåÅÂ§ßÈáèËßÑÂàôÔºâ
            ('grpc.max_send_message_length', 50 * 1024 * 1024),  # 50MB
            ('grpc.max_receive_message_length', 50 * 1024 * 1024),  # 50MB
        ]
        
        with grpc.insecure_channel(self.address, options=options) as channel:
            stub = bazi_rule_pb2_grpc.BaziRuleServiceStub(channel)
            try:
                response = stub.MatchRules(request, timeout=self.timeout)
                
                import datetime
                response_time = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
                matched_data = json.loads(response.matched_json) if response.matched_json else []
                matched_count = len(matched_data)
                
                # Â§ÑÁêÜ unmatched Êï∞ÊçÆÔºàÂèØËÉΩÊòØÂÆåÊï¥ÂàóË°®ÊàñÂè™ÂåÖÂê´ countÔºâ
                unmatched_json = response.unmatched_json if response.unmatched_json else '{}'
                unmatched_data = json.loads(unmatched_json)
                if isinstance(unmatched_data, dict) and 'count' in unmatched_data:
                    # Âè™ËøîÂõû countÔºå‰∏çËøîÂõûÂÆåÊï¥ÂàóË°®
                    unmatched_list = []
                else:
                    unmatched_list = unmatched_data if isinstance(unmatched_data, list) else []
                
                logger.info(f"[{response_time}] ‚úÖ bazi-rule-service (gRPC): Ë∞ÉÁî®ÊàêÂäüÔºåÂåπÈÖç {matched_count} Êù°ËßÑÂàô")
                return {
                    "matched": matched_data,
                    "unmatched": unmatched_list,
                    "context": json.loads(response.context_json) if response.context_json else {},
                }

            except grpc.RpcError as e:
                import datetime
                error_time = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
                logger.error(f"[{error_time}] ‚ùå bazi-rule-service (gRPC): Ë∞ÉÁî®Â§±Ë¥• - {e}")
                raise

    def health_check(self) -> bool:
        """ÂÅ•Â∫∑Ê£ÄÊü•"""
        request = bazi_rule_pb2.HealthCheckRequest()
        try:
            with grpc.insecure_channel(self.address) as channel:
                stub = bazi_rule_pb2_grpc.BaziRuleServiceStub(channel)
                response = stub.HealthCheck(request, timeout=5.0)
                return response.status == "ok"
        except grpc.RpcError:
            logger.exception("bazi-rule-service health check failed")
            return False

